base/compiler/ssair/ir.jl

```diff
@@ -1595,3 +1605,373 @@ function iterate(x::BBIdxIter, (idx, bb)::Tuple{Int, Int}=(1, 1))
+            else
+                1
+            end for ibb in 1:length(ir.cfg.blocks)
+        ],
+    )
+    newblocks = 2 * length(statement_positions)
+
+    # Insert `newblocks` new blocks:
+    oldnblocks = length(ir.cfg.blocks)
+    resize!(ir.cfg.blocks, oldnblocks + newblocks)
+    # Copy pre-existing blocks:
+    for iold in oldnblocks:-1:1
+        bb = ir.cfg.blocks[iold]
+        for labels in (bb.preds, bb.succs)
+            for (i, l) in pairs(labels)
+                labels[i] = bbchangemap[l]
+            end
+        end
+        start = ssachangemap[bb.stmts.start]
+        stop = ssachangemap[bb.stmts.stop]
+        ir.cfg.blocks[bbchangemap[iold]] = BasicBlock(bb, StmtRange(start, stop))
+    end
+    # Insert new blocks:
+    for iold in target_blocks
+        positions = block_to_positions[iold]
+        ilst = bbchangemap[iold]  # using bbchangemap as it's already moved
+        bblst = ir.cfg.blocks[ilst]
+
+        inew = get(bbchangemap, iold - 1, 0)
+        preoldpos = 0  # for detecting duplicated positions
+        p1 = first(bblst.stmts)
+        isfirst = true
+        for (i, oldpos) in pairs(positions)
+            if preoldpos == oldpos
+                p2 = p1
+            else
+                preoldpos = oldpos
+                p2 = get(ssachangemap, oldpos - 1, 0) + 1
+                if isfirst
+                    isfirst = false
+                    p1 = min(p1, p2)
+                end
+            end
+            p3 = p2 + 1
+            @assert p1 <= p2 < p3
+            ir.cfg.blocks[inew+1] = BasicBlock(StmtRange(p1, p2))
+            ir.cfg.blocks[inew+2] = BasicBlock(StmtRange(p3, p3))
+            p1 = p3 + 1
+            inew += 2
+        end
+        ifst = get(bbchangemap, iold - 1, 0) + 1
+        bbfst = ir.cfg.blocks[ifst]
+        for p in bblst.preds
+            k = findfirst(==(ilst), ir.cfg.blocks[p].succs)
+            @assert k !== nothing
+            ir.cfg.blocks[p].succs[k] = ifst
+        end
+        copy!(bbfst.preds, bblst.preds)
+        empty!(bblst.preds)
+        stmts = StmtRange(ssachangemap[positions[end]], last(bblst.stmts))
+        ir.cfg.blocks[bbchangemap[iold]] = BasicBlock(bblst, stmts)
+        @assert !isempty(stmts)
+    end
+    for bb in ir.cfg.blocks
+        @assert !isempty(bb.stmts)
+    end
+    cfg_reindex!(ir.cfg)
+
+    # Like `bbchangemap` but maps to the first added BB (not the last)
+    gotolabelchangemap = _cumsum!(
+        Int[
+            if ibb in target_blocks
+                1 + 2 * length(block_to_positions[ibb])
+            else
+                1
+            end for ibb in 0:length(ir.cfg.blocks)-1
+        ],
+    )
+
+    on_ssavalue(v) = SSAValue(ssachangemap[v.id])
+    on_phi_label(l) = bbchangemap[l]
+    on_goto_label(l) = gotolabelchangemap[l]
+    for stmts in (ir.stmts, ir.new_nodes.stmts)
+        for i in 1:length(stmts)
+            st = stmts[i]
+            inst = ssamap(on_ssavalue, st[:inst])
```

Comment:
Since `userefs` doesn't yet count SSAValues themselves as uses (pending me landing #44794), `on_ssavalue` will not be applied to a `st[:inst]::SSAValue`. They should probably be handled here until that functionality changes

Reply:
Thanks for catching this! I had this in the original code but I missed it while re-writing this patch using `userefs`.

Reply:
@tkf #44794 is merged now so this change should no longer be necessary!

Reply:
Thanks, that's the best kind of review! :)